datalad/cmd2.py

```diff
@@ -0,0 +1,86 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+Wrapper for command calls, allowing for dry runs and output handling
+
+"""
+
+
+import subprocess
+import shlex
+import sys
+import logging
+
+lgr = logging.getLogger('datalad.cmd')
+
+
+class Runner(object):
+    """Provides a wrapper for system calls.
+
+    An object of this class provides a method to make system calls.
+    Outputs (stdout and stderr) are streamed to stdout during execution as if you were calling it from command line.
+    Additionally allows for dry runs. This is achieved by initializing the `Runner` with `dry=True`.
+    The Runner will then collect all calls as strings in `cmdBuffer`.
+    """
+
+    __slots__ = ['cmdBuffer', 'dry']
+
+    def __init__(self, dry=False):
+        self.dry = dry
+        self.cmdBuffer = []
+
+    def run(self, cmd):
+        """
+
+        Parameters
+        ----------
+        cmd : str
+          String defining the command call.
+
+        Returns
+        -------
+        Status code as returned by the called command.
+        """
+
+        self.log("Running: %s" % cmd)
+
+        if not self.dry:
+            # pipeArgs = shlex.split(cmd)
+
+            proc = subprocess.Popen(cmd, stdout=sys.stdout, stderr=subprocess.STDOUT, shell=True)
```

Comment:
>      anyways -- how does this improve upon existing solution??
>      The intended difference is the use of subprocess and live output of
>      stdout as discussed.

>      I see no drycall which allowed for dry execution/reporting of python
>      native functions
>      I would have preserved it, when replacing the old file. But how do you
>      want to use it?

the way it  is already used:

$> git grep drycall         
datalad/cmd.py:    def drycall(self, f, *args, **kwargs):
datalad/crawler/main.py:        _call = self.runner.drycall
datalad/support/repos.py:            self.runner.drycall(self.write_description, description)
datalad/support/repos.py:                self.runner.drycall(os.unlink, full_filename)
datalad/support/repos.py:    _call = runner.drycall


ie to log the intent of some actions to be taken

-- 
Yaroslav O. Halchenko, Ph.D.
http://neuro.debian.net http://www.pymvpa.org http://www.fail2ban.org
Research Scientist,            Psychological and Brain Sciences Dept.
Dartmouth College, 419 Moore Hall, Hinman Box 6207, Hanover, NH 03755
Phone: +1 (603) 646-9834                       Fax: +1 (603) 646-1419
WWW:   http://www.linkedin.com/in/yarik